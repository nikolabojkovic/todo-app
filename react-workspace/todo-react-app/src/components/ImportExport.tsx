import { faFileExport, faFileImport } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useRef, useState } from "react";
import { Button, Col, Container, Form, InputGroup, Row } from "react-bootstrap";
import { useTodoList, useTodoListDispatch } from "../context/TodosContext";
import { Todo } from "../models/Todo";

const buttonStyle : any = { 
  backgroundColor: '#F5F6F7', 
  borderRadius: '20px', 
  minWidth: '90px'
}

const fileReader = new FileReader();

export function ImportExport() {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();
  
  const [file, setFile] = useState(null);
  const fileRef = useRef(null as any);

  function handleImport(e: any) {
    e.preventDefault();

    if (!file) 
      return;

    fileReader.readAsText(file);
  };

  function handleExport() {
    const link = document.createElement('a');
    const jsonContent = `data:text/json;chatset=utf-8,${encodeURIComponent(
      JSON.stringify(todoList.originalList)
    )}`;
    link.href = jsonContent;
    link.download = `todo-list-${new Date().toLocaleDateString()}-${new Date().toLocaleTimeString()}.json`;
    link.click();
  }

  fileReader.onload = function (e: any) {
    const text = e.target.result;
    const list = JSON.parse(text) as Todo[];

    if (!(list instanceof Array)) {
      alert("Invalid JSON file content. Todo list should be an array.");
      return;
    }

    const importedTodoList = list.map((item: any) => new Todo(item.id, 
                                                              item.title, 
                                                              item.description, 
                                                              item.completed, 
                                                              item.createdAt));

    if ((importedTodoList.length > 0 
        && (!(importedTodoList[0] instanceof Todo) 
         || !(Todo.validateFields(importedTodoList[0]))
            )
        )) {
      alert("Invalid JSON file content. Objects in array are not valid Todo objects.");
      return;
    }
          
    setFile(null);
    dispatch({
      type: 'imported',
      originalList: importedTodoList,
      activePage: 1
    });
    fileRef.current.value = '';
  };

  return (
    <Form className="todo-background p-1">
      <Container fluid>
        <Row xs={1} sm={3}>
          <Col sm={8} className="p-2">
            <InputGroup size="sm" className="">
              <Form.Control
                aria-label="Choose File"
                aria-describedby="inputGroup-sizing-sm"
                type={"file"} 
                accept={".json"} 
                ref={fileRef}
                onChange={(e: any) => {
                  setFile(e.target.files[0]);
                }}
                readOnly 
              />
            </InputGroup>
          </Col>
          <Col sm={2} className="p-2">
            <Button 
              className="w-100"
              variant="outline-success" 
              size="sm"
              style={ buttonStyle }
              onClick={handleImport}
              disabled={!file}
            >
              Import {' '}
              <FontAwesomeIcon icon={faFileImport} />
            </Button>
          </Col>
          <Col sm={2} className="p-2">
            <Button 
              className="w-100"
              variant="outline-success" 
              size="sm"
              style={ buttonStyle }
              disabled={!todoList.originalList || todoList.originalList.length === 0}
              onClick={handleExport}
            >
              Export {' '}
              <FontAwesomeIcon icon={faFileExport} />
            </Button>
          </Col>
        </Row>
      </Container>
    </Form>
  )
}