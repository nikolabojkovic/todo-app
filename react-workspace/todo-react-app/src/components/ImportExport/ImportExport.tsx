import { faFileExport, faFileImport } from "@fortawesome/free-solid-svg-icons";
import { Button, Col, Container, Form, InputGroup, Row } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { ChangeEvent, useRef, useState } from "react";

import { useTodoList, useTodoListDispatch } from "../../context";
import { ConfirmModal } from "../ConfirmModal/ConfirmModal";
import { ITodo, Todo, IAction, TodoActions } from "../../models";

type Props = {
  downloadLink: HTMLAnchorElement,
  fileReader: FileReader,
  alert: (message: string) => void
}

export function ImportExport({ downloadLink, fileReader, alert }: Props) {
  const todoList = useTodoList();  
  const dispatch = useTodoListDispatch();
  
  const [showModal, setShowModal] = useState(false);
  const [file, setFile] = useState<File | null>(null);
  const fileRef = useRef(null as HTMLInputElement | null);

  function handleExport() {
    const jsonContent = `data:text/json;chatset=utf-8,${encodeURIComponent(
      JSON.stringify(todoList.originalList)
    )}`;
    downloadLink.href = jsonContent;
    downloadLink.download = `todo-list-${new Date().toLocaleDateString()}-${new Date().toLocaleTimeString()}.json`;
    downloadLink.click();
  }

  function handleImport() {
    fileReader.readAsText(file!);          
    setFile(null);
    fileRef!.current!.value = '';
  }

  fileReader.onload = handleFileContent;

  function handleFileContent(e: ProgressEvent<FileReader>) {
    const text = e.target?.result;
    const list = JSON.parse(text as string) as Todo[];

    if (!(list instanceof Array)) {
      alert('Invalid JSON file content. Todo list should be an array.');
      return;
    }

    const importedTodoList = list.map((item: ITodo) => new Todo(item.id, 
                                                                item.title, 
                                                                item.description, 
                                                                item.completed, 
                                                                item.createdAt,
                                                                item.sortId));

    if ((importedTodoList.length > 0 
        && (!(importedTodoList[0] instanceof Todo) 
         || !(Todo.validateFields(importedTodoList[0]))
            )
        )) {
      alert('Invalid JSON file content. Objects in array are not valid Todo objects.');
      return;
    }          
    
    dispatch({
      type: TodoActions.imported,
      payload: {
        list: importedTodoList,
        activePage: 1
      }
    } as IAction);
  }

  return (
    <>
      <Form className="todo-background p-1">
        <Container fluid>
          <Row xs={1} sm={3}>
            <Col sm={8} className="p-2">
              <InputGroup size="sm" className="">
                <Form.Control
                  data-testid="choose-file"
                  aria-label="Choose File"
                  aria-describedby="inputGroup-sizing-sm"
                  type={"file"} 
                  accept={".json"} 
                  ref={fileRef}
                  onChange={(e: ChangeEvent<HTMLInputElement>) => {
                    setFile(((e.target).files![0]));
                  }}
                  readOnly 
                />
              </InputGroup>
            </Col>
            <Col sm={2} className="p-2">
              <Button 
                data-testid="import-button"
                variant="outline-secondary"
                className="w-100 action-button"
                size="sm"
                onClick={() => { 
                  if (todoList.settings.general.isConfirmEnabled) {
                    setShowModal(true);

                    return;
                  }

                  handleImport();
                }}
                disabled={!file}
              >
                Import {' '}
                <FontAwesomeIcon icon={faFileImport} />
              </Button>
            </Col>
            <Col sm={2} className="p-2">
              <Button 
                data-testid="export-button"
                variant="outline-secondary"
                className="w-100 action-button"
                size="sm"
                disabled={!todoList.originalList || todoList.originalList.length === 0}
                onClick={() => handleExport()}
              >
                Export {' '}
                <FontAwesomeIcon icon={faFileExport} />
              </Button>
            </Col>
          </Row>
        </Container>
      </Form>
      <ConfirmModal 
        content={'Existing data will be lost. Are you sure?'} 
        show={showModal}
        onConfirm={() => {          
          setShowModal(false);
          handleImport();
        }}
        onCancel={() => setShowModal(false)}
      />
    </>
  );
}