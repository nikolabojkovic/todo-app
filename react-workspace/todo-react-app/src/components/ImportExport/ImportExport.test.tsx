import renderer from 'react-test-renderer';
import { ImportExport } from './ImportExport';

describe('ImportExport', () => {
  it('component should match snapshot', () => {
    const jsxElement = <ImportExport />;
    const tree = renderer.create(jsxElement).toJSON();
    expect(tree).toMatchSnapshot();
  });
});