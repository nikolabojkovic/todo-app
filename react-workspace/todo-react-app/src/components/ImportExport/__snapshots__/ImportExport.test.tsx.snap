// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportExport component should match snapshot 1`] = `
<form
  className="todo-background p-1"
>
  <div
    className="container-fluid"
  >
    <div
      className="row row-cols-sm-3 row-cols-1"
    >
      <div
        className="p-2 col-sm-8"
      >
        <div
          className="input-group input-group-sm"
        >
          <input
            accept=".json"
            aria-describedby="inputGroup-sizing-sm"
            aria-label="Choose File"
            className="form-control"
            onChange={[Function]}
            readOnly={true}
            type="file"
          />
        </div>
      </div>
      <div
        className="p-2 col-sm-2"
      >
        <button
          className="w-100 action-button btn btn-outline-secondary btn-sm"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Import 
           
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-file-import "
            data-icon="file-import"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128 64c0-35.3 28.7-64 64-64H352V128c0 17.7 14.3 32 32 32H512V448c0 35.3-28.7 64-64 64H192c-35.3 0-64-28.7-64-64V336H302.1l-39 39c-9.4 9.4-9.4 24.6 0 33.9s24.6 9.4 33.9 0l80-80c9.4-9.4 9.4-24.6 0-33.9l-80-80c-9.4-9.4-24.6-9.4-33.9 0s-9.4 24.6 0 33.9l39 39H128V64zm0 224v48H24c-13.3 0-24-10.7-24-24s10.7-24 24-24H128zM512 128H384V0L512 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
      <div
        className="p-2 col-sm-2"
      >
        <button
          className="w-100 action-button btn btn-outline-secondary btn-sm"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Export 
           
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-file-export "
            data-icon="file-export"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 64C0 28.7 28.7 0 64 0H224V128c0 17.7 14.3 32 32 32H384V288H216c-13.3 0-24 10.7-24 24s10.7 24 24 24H384V448c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V64zM384 336V288H494.1l-39-39c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l80 80c9.4 9.4 9.4 24.6 0 33.9l-80 80c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l39-39H384zm0-208H256V0L384 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</form>
`;
